### PR TITLE
refactor: change `PoissonLikelihood` into a function with faster compilation, speed and low memory footprint

### PR DESCRIPTION
**Pull request by Qazalbash**
## Summary

We have changed `PoissonLikelihood` into `poisson_likelihood,` which compiles very fast and takes less memory than before.

## Related To

- #454 

## Description

In this PR, we have explored the use **padded arrays** for data to make their shapes same. We padded the data, reference priors, and an extra mask array to remember the padded values. When we have the same shape arrays, we could apply the `jax.vmap` transformation to calculate the likelihood of a single event. This was exhausting resources, therefore we moved to `jax.lax.scan` and it is working like a charm. We are getting faster compilation times, good speed, and low memory consumption.

### CODE REVIEW

**Comment by zeeshan5885** [comment]
This change significantly improves speed and memory consumption. The best optimization so far. :-)

File: src/gwkokab/inference/poissonlikelihood.py [hunk 1]
```diff
@@ -176,3 +179,127 @@ def log_posterior(self, x: Array, _: dict) -> Array:
...
+    (data,) = promote_args_inexact("segment_logsumexp", data)
+
+    amax = jnp.max(data.real, initial=-jnp.inf)
+    amax = lax.stop_gradient(
+        lax.select(jnp.isfinite(amax), amax, lax.full_like(amax, 0))
+    )
+
+    exp_a = lax.exp(lax.sub(data, amax))
+
+    sumexp = jax.ops.segment_sum(
```
[the remark this answers is not in the document]
Comment (Qazalbash) [reply src/gwkokab/inference/poissonlikelihood.py:202]: These lines will be removed in future commits.

**Review by zeeshan5885 — APPROVED** [review]
Great Optimization Meesum. I also did stress test, it worked great.
